taf/repository_tool.py

```diff
@@ -173,31 +261,157 @@ def set_metadata_expiration_date(self, role, start_date=datetime.datetime.now(),
+      - MetadataUpdateError: If any other error happened during metadata update
     """
-    Generates metadata of just one metadata file, corresponding
-    to the provided role
-    args:
-      repository: a tu repository
-      role: a tuf role
-      keystore: location of the keystore file
-      update_snapshot_and_timestamp: should timestamp and snapshot.json also be updated
+    try:
+      snapshot_date = kwargs.get('snapshot_date', datetime.datetime.now())
+      snapshot_interval = kwargs.get('snapshot_interval', None)
+
+      timestamp_date = kwargs.get('timestamp_date', datetime.datetime.now())
+      timestamp_interval = kwargs.get('timestamp_interval', None)
+
+      snapshot_key = load_role_key(keystore, 'snapshot')
+      self.update_snapshot(snapshot_key, snapshot_date, snapshot_interval, write=False)
+
+      timestamp_key = load_role_key(keystore, 'timestamp')
+      self.update_timestamp(timestamp_key, timestamp_date, timestamp_interval, write=False)
+
+      self._repository.writeall()
+
+    except (TUFError, SSLibError) as e:
+      raise MetadataUpdateError('all', str(e))
+
+  def update_snapshot(self, snapshot_key, start_date=datetime.datetime.now(), interval=None, write=True):
+    """Update snapshot metadata.
+
+    Args:
+      - snapshot_key(securesystemslib.formats.RSAKEY_SCHEMA): Snapshot key.
+      - start_date(datetime): Date to which the specified interval is added when
+                              calculating expiration date. If a value is not
+                              provided, it is set to the current time
+      - interval(int): A number of days added to the start date. If not provided,
+                      the default value is used
+      - write(bool): If True snapshot metadata will be signed and written
+
+    Returns:
+      None
+
+    Raises:
+      - InvalidKeyError: If wrong key is used to sign metadata
+      - SnapshotMetadataUpdateError: If any other error happened during metadata update
     """
-    private_role_key = load_role_key(role, keystore)
-    self._role_obj(role).load_signing_key(private_role_key)
-    # only write this role's metadata
-
-    if not update_snapshot_and_timestamp:
-      self._repository.write(role)
-    else:
-      snapshot_key = load_role_key('snapshot', keystore)
+    from .sc_utils import is_valid_metadata_key
+
+    try:
+      if not is_valid_metadata_key(self, 'snapshot', snapshot_key):
+        raise InvalidKeyError('targets')
+
+      self.set_metadata_expiration_date('snapshot', start_date, interval)
+
       self._repository.snapshot.load_signing_key(snapshot_key)
-      timestamp_key = load_role_key('timestamp', keystore)
+      if write:
+        self._repository.write('snapshot')
+
+    except (SmartCardError, TUFError, SSLibError) as e:
+      raise SnapshotMetadataUpdateError(str(e))
+
+  def update_timestamp(self, timestamp_key, start_date=datetime.datetime.now(), interval=None, write=True):
```

Comment:
We had something similar.
There are few reasons I have split it to different methods:

1. easier to read/search when using as specific method
2. less space to make an error when calling (not passing a string value for role anymore)
3. different way of signing can be introduced (like for targets)

I could add a helper method, but I think code readability will suffer and it will be used just inside those two methods. If we introduce delegation roles, that would probably make sense

Reply:
I have added two helper methods, but kept the specific ones.

Reply:
That's what I had in mind